### ISSUE TITLE
using vcpkg release static libraries in Visual Studio debug builds fails

### ISSUE THREAD

**Issue by BeErikk** [Microsoft/vcpkg#2153]
If I understand it right, the idea is that you should use _debug_ versions of your _vcpkg_ builds when you are doing Visual Studio debug builds of your project. It's a different approach compared with for example the _Windows SDK_ where all libraries are release builds. I guess it can be useful in some circumstances, but it surely adds complexity. I think if I add the `-nodefaultlib` linker option and the `-Zl` compiler option to vcpkg _static library_ builds, I have a choice to do either way. If so, how can I add these options in a convenient way? 

edit:
There's also the issue of mixing vcpkg static library with dynamic linked libraries. A thing you most certainly want to do.

**Comment by BeErikk** [comment]
Just to add some depth: My current problem is I need static `zlib`, static `curl` and dynamic `openssl`. This work fine with my own compiles, but I do want to use vcpkg builds instead.

I'm looking at curl `portfile.cmake`. How and what can I add to get the following in the cmake cache
```
//Flags used by the linker during the creation of static libraries.
CMAKE_STATIC_LINKER_FLAGS:STRING=-machine:x64 -debug -nodefaultlib
```
AFAIK there is no entry for compiler flags for static builds, so how to add `/Zl (Omit Default Library Name)` to `CMAKE_C_FLAGS` just for static builds?

**Comment by ras0219-msft** [comment]
Unfortunately, the MSVC CRT is not compatible between Debug and Release. For example, it is unsupported to `malloc()`/`new` in the Release CRT but then call `free()`/`delete` from the Debug CRT.
 The Windows SDK jumps through many hoops to avoid this causing a problem, but for most libraries you _must_ use debug versions when building your application as debug (and same for release).

>My current problem is I need static zlib, static curl and dynamic openssl. This work fine with my own compiles, but I do want to use vcpkg builds instead.

I'd recommend achieving this by making a custom triplet; see our documentation[1] under "Per-port customization" for more info on that.

>how to add `/Zl (Omit Default Library Name)` to `CMAKE_C_FLAGS` just for static builds?

For CFLAGS, you can set `VCPKG_C_FLAGS` inside the triplet file. However, it looks like our support for linker flags is incomplete at the moment. If the above doesn't help, we can look at plumbing the linker flags through! (However, note that at the moment OpenSSL may not obey the `C_FLAGS` setting because it uses a custom perl buildsystem.)

[1] https://vcpkg.readthedocs.io/en/latest/users/triplets/ or https://github.com/Microsoft/vcpkg/blob/master/docs/users/triplets.md

**Comment by BeErikk** [comment]
Thanks @ras0219-msft
Regarding debug/release versions of msvcrt, I thought the `-Zl` switch stripped the library from any external dependencies and kept the compiled source as is, put the dependencies "on hold" so to speak and to be resolved when the library is being used. This may not be a completely supported way though.

One way is to follow the guidelines in makefile settings of the DDK build Tool build.exe (latest Windows 7 DDK), which I believe used to be the tool to build the Windows SDK. For a release build these settings are
`-Oxs -Oy -Zel -Zp8 -Gy -Gi- -Gm- -W4 -WX -Gz -G6 -GX- -GR- -GF -GS -Z7 `
and the linker settings always specify `-nodefaultlib`. Some options may be obsolete.
If someone is interested, here is an old log from building bzip2 with DDK build :D
[buildfre_wnet_x86.log](https://github.com/Microsoft/vcpkg/files/1454691/buildfre_wnet_x86.log)

Anyway, earlier I tried the following patch in curl's `port.cmake`

```
SET(NODEFAULTLIB_COMPILE_FLAGS "-Zl")
SET(NODEFAULTLIB_LINK_FLAGS "-nodefaultlib")
SET( CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} ${NODEFAULTLIB_LINK_FLAGS}" )

if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
    SET(CURL_STATICLIB OFF)
else()
    SET(CURL_STATICLIB ON)
    SET( CMAKE_C_FLAGS  "${CMAKE_C_FLAGS} ${NODEFAULTLIB_COMPILE_FLAGS}" )
endif()

```
then removed all curl installs, deleted the curl buildtree and reinstalled curl
`vcpkg install curl:x64-windows-static curl:x64-windows`

Unfortunately no trace of `nodefaultlib` in the buildtree

I will try with your suggestions and see if I'm more lucky.

**Comment by BeErikk** [comment]
Apparently I should start with learning how to read. **The triplet approach  is brilliant**!

I made the following triplet
`x64-windows-static-dyncrt.cmake`
```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
SET(NODEFAULTLIB_COMPILE_FLAGS "-Zl")
SET(NODEFAULTLIB_LINK_FLAGS "-nodefaultlib")
SET(VCPKG_C_FLAGS  "${VCPKG_C_FLAGS} ${NODEFAULTLIB_COMPILE_FLAGS}")
SET(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} ${NODEFAULTLIB_LINK_FLAGS}" )
```
starting all over with the new triplet 
`vcpkg install zlib:x64-windows-static-dyncrt`
gives an error
[triplet-error.txt](https://github.com/Microsoft/vcpkg/files/1454932/triplet-error.txt)
failing this line
`file(MAKE_DIRECTORY ${CURRENT_BUILDTREES_DIR} ${CURRENT_PACKAGES_DIR})`

**Comment by ras0219-msft** [comment]
That looks like it might be a permissions or disk space error; could you check those first? Then, try deleting the entire `buildtrees\` folder and rebuilding.

**Comment by BeErikk** [comment]
no after delete of buildtrees, vcpkg downloaded and created the zlib buildtrees/src as expected,
but failed` vcpkg install zlib:x64-windows-static-dyncrt`
`vcpkg install curl:x64-windows` succeeds though

I'm uncertain on the syntax in the triplet, is it OK? Also, the last line may not work

**Comment by ras0219-msft** [comment]
Let's start off simple -- can you confirm that this works:
```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
```

Then, the first thing you wanted was to link openssl dynamically
```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
if(PORT STREQUAL "openssl")
  set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
```

**Comment by BeErikk** [comment]
Yes, the first triplet worked - **which gives there's something wrong with my syntax.**
Thinking about the second triplet and it's implications. Always having openssl as dlls makes sense I think.  Testing it will take some time. At least the first triplet will give the benefit that all static packages default to dynamic CRT.

**Comment by BeErikk** [comment]
Just a short update: I spent a good deal of time hunting errors including recompiling and debugging vcpkg, until I found the cause. I was running vcpkg from a Visual Studio 2017 prompt - `x64 Native Tools Command Prompt for VS 2017` which somehow interfered. Running my triplet from an ordinary command console work as expected. 

However, applying `-Zl` straight to `VCPKG_C_FLAGS` makes cmake config fail of course, since all dependencies are stripped. I would need something like a `VCPKG_C_FLAGS_STATIC` variable.
